Timeout and retry `xcodebuild -showBuildSettings`

fcy opened the pull request:
As it may hang forever.

Fixes #4059

Commit:
Timeout and retry 'xcodebuild -showBuildSettings'

Update run_command to use Helper.backticks, and add the ability to retry before raising Timeout::Error

Fixes #4059

fcy commented:
@asfalcone FYI

KrauseFx commented on fastlane_core/lib/fastlane_core/project.rb:
```diff
@@ -303,5 +303,67 @@ def parsed_info
       !!ENV["AUTOMATED_SCHEME_SELECTION"]
     end
+
+    def backticks(command, print: true, timeout: false, retries: 0)
```
This seems like a duplicate of what we have in `fastlane_core`, right?

fcy replied:
Do you mean the `Helper.backticks`? If yes, this is a private version that adds timeout and retry mainly because in [this PR](https://github.com/fastlane/fastlane/pull/4492) it was suggested that having this locally to the one spot it is currently needed would be better.

I think the approach there is better because it adds this functionality to whomever needs it but I decided to open a simplified version to discuss it.

fcy replied:
I have renamed this method to avoid the confusion.

fcy replied:
Oh actually I think you referring to `Project.run_command` which has timeout but can't retry. I'll refactor this.

fcy commented:
From this [comment](https://github.com/fastlane/fastlane/pull/5188#discussion_r68113906):

> Oh actually I think you referring to Project.run_command which has timeout but can't retry. I'll refactor this.

My proposal there is to add the functionality to retry to `run_command`, so move the implementation of the private `retryable_exec` to `run_command` with a default timeout of zero and default retry of zero too. Another benefit is that it would stop using [the dangerous `Timeout`](http://jvns.ca/blog/2015/11/27/why-rubys-timeout-is-dangerous-and-thread-dot-raise-is-terrifying/).

I'll make these changes if you guys think this is the correct approach.

KrauseFx commented:
I'm not a big fan of having this extra method just for this. It feels like we're not fixing the underlying error, but just try to work around it. I'd love to find a cleaner solution that works for everyone 👍

fcy commented:
Just to be sure, is the problem the private method that is only used for the `-showBuildSettings`? Or the fact that `xcodebuild` works when we retry to same command?

I can fix the first one by adding support to retry to `run_command`, and two more environment variables to customize it: `FASTLANE_XCODE_BUILD_SETTINGS_TIMEOUT`, and `FASTLANE_XCODE_BUILD_SETTINGS_RETRIES`, just like the existing `FASTLANE_XCODE_LIST_TIMEOUT`.

The second problem is about deciding when Fastlane tries to work around Xcode and when it gives up. In my opinion part of what makes Fastlane so amazing is how reliable and easy it makes using Xcode from the command line.

I'd love to contribute more, just let me know in clear words what is expected. I won't be offended or anything like it by any of the choices.

neonichu commented:
@fcy it would be great if you could implement `retryable_exec` so that it wraps the current `Helper.backticks` functionality, instead of duplicating it.

fcy commented:
@neonichu @KrauseFx please take another look. I update `run_command` to also use the existing `Helper.backticks` and added the ability to retry. Along with two new environment variables to control timeout and number of retries for the `xcodebuild -showBuildSettings`.

neonichu commented:
LGTM, could you update the environment vars to something like `FASTLANE_XCODEBUILD_SETTINGS_TIMEOUT`, just making `XCODEBUILD` one word.

fcy commented:
I definitely can.

But for some reason that I still don't understand this new version this work as well as the old one. I'm using it in my CI and the build started hanging on `xcodebuild -list` which wasn't even a problem before. What is even more strange is that `xcodebuild -list` already was in the `Timeout.timeout` function but it is not timing out anymore. The only difference inside the `Timeout.timeout` block is that it uses `Helper.backticks` instead of `` `command` ``.

I'm going to make the changes to the env vars, rebase, and try reverting the `timeout` back to just to see if it is related.

Commit:
Rename env var from XCODE_BUILD to XCODEBUILD

Revert run_command to use backticks again instead of the Helper, because using the Helper seem to make timeout not behaving as nicely.

Commit:
Change default to retry 3 times

The original implementation used up to 3 times and it worked flawlessly. I changed it to 1 hoping it would be enough but it isn't.

fcy commented:
Changes are done and it is battled tested in our CI, it works as flawlessly as PR #4492.

`Timeout.timeout` does not work when using `Helper.backticks` it just doesn't timeout and let the build hanging. Reverting back to `` `#{command}` `` fixed the issue and it now times-out and retries as expected.

Let me know if you prefer that I squashed the 3 commits before merging.

fcy commented:
Anything else I could do here? This has been running on our CI for the past week without any problems, it solved the issue.

Commit:
Merge branch 'master' into fcy/fix-xcodebuild-show-settings

# Conflicts:
#	fastlane_core/lib/fastlane_core/project.rb

fcy committed:
Print -showBuildSettings command

Omitting it causes confusion on what command is hanging.

fcy committed:
Fix silent variable

larslockefeer commented:
We were seeing the same issue, [applying the patch of this PR on top of the currently released fastlane version](https://github.com/larslockefeer/fastlane/tree/travis-fix) solved the issue for us as well.

andyshephard commented:
Will this be merged anytime soon? Experiencing issue w/ Bitrise CI that this PR would fix (potentially)

Commit:
Merge branch 'master' into fcy/fix-xcodebuild-show-settings

fcy commented:
I have merged master into here to update it. In case other are using this branch I'm not rewriting the commit history with rebases.

Commit:
Merge branch 'master' into fcy/fix-xcodebuild-show-settings

Commit:
Make Project.run_command only returns the output

Callers are responsible to store the output on their own variables.

revile commented:
Any progress on getting this merged? Running tests via fastlane is _very_ unreliable at the moment.

KrauseFx commented:
Hey everyone, really sorry for the delay on this. As you know, we can't reliably reproduce this issue unfortunately, however I'll take a look at the code, and make sure everything makes sense, has enough documentation, and doesn't break anything 👍

KrauseFx commented on fastlane_core/lib/fastlane_core/project.rb:
```diff
@@ -301,13 +312,33 @@ def self.xcode_list_timeout
     # @returns the output
-    # Note: currently affected by fastlane/fastlane_core#102
-    def self.run_command(command, timeout: 0)
+    # Note: - currently affected by fastlane/fastlane_core#102
```
This was probably migrated to fastlane/fastlane?

KrauseFx commented on fastlane_core/lib/fastlane_core/project.rb:
```diff
@@ -301,13 +312,33 @@ def self.xcode_list_timeout
-    # Note: currently affected by fastlane/fastlane_core#102
-    def self.run_command(command, timeout: 0)
+    # Note: - currently affected by fastlane/fastlane_core#102
+    #       - retry feature added to solve fastlane#4059
```
Can you include the full URLs here instead?

KrauseFx commented on fastlane_core/lib/fastlane_core/project.rb:
```diff
@@ -301,13 +312,33 @@ def self.xcode_list_timeout
+
+      tries = 1
+      begin
+        return Timeout.timeout(timeout) { `#{command}`.to_s }
```
Could we call `Helper.backticks` here instead? This way the command and its result will be printed out to the user

KrauseFx replied:
Let's also change this to be multi line, using `do .. end`

KrauseFx replied:
And you can add `self.xcodebuild_list_silent` here too, by using `Helper.backticks(command, print: !self.xcodebuild_list_silent)`

fcy replied:
My original implementation used it, on unit tests it worked just fine but in real life using `Helper.backticks` made the `Timeout` useless. It just don't retry nor times out ever, it is like the `Timeout.timeout` call didn't even exist.

KrauseFx replied:
But `Helper.backticks` is just using backticks: https://github.com/fastlane/fastlane/blob/master/fastlane_core/lib/fastlane_core/helper.rb#L15-L20

fcy replied:
Yes I noticed, I can't say I know enough Ruby to understand it. Feels totally "wat?". 

Everything I read said to just not use `Timeout` (links in older comments here) but that means not using backticks. Actually not using `Timeout` was my first implementation, which was based on #4492. I changed it just because the core team didn't seem to agree with that approach.

KrauseFx replied:
Alright, if you say it doesn't work with Helper.backticks, I'm ok with leaving it like this, however please do print out the command and the command output, similar to https://github.com/fastlane/fastlane/blob/master/fastlane_core/lib/fastlane_core/helper.rb#L15-L20

How did you test to see that this really doesn't work?

fcy replied:
I will 👍 

---

I use Bundler to manage the tools we use (CocoaPods, Fastlane, and Danger), basically I just put the git branch I want to use for Fastlane. I was successfully using $4492 when I decided to create a simplified PR using the same implementation it did but totally compartmentalized to `project.rb` this is the progress of changes:

1. #4492, worked
2. #4492 applied to `project.rb`, worked
3. [early feedback](https://github.com/fastlane/fastlane/pull/5188#issuecomment-229654362) asked to use `Helper.backticks`, retries stopped working and the build hand just like Fastene in master
4. Went back to the original backticks in `project.rb` instead of `Helper.backticks`, and it worked again (literally nothing else changed)

All these I put to run on our CI which makes the problem happen more often (probably because of the many times the build runs).

KrauseFx commented on fastlane_core/lib/fastlane_core/project.rb:
```diff
@@ -301,13 +312,33 @@ def self.xcode_list_timeout
+      begin
+        return Timeout.timeout(timeout) { `#{command}`.to_s }
+      rescue Timeout::Error
+        raise if tries >= retries
```
Let's add the following here
```ruby
UI.verbose("Command timed out, trying again...")
```

KrauseFx commented on fastlane_core/lib/fastlane_core/project.rb:
```diff
@@ -216,7 +216,18 @@ def build_xcodebuild_showbuildsettings_command
       unless @build_settings
         command = build_xcodebuild_showbuildsettings_command
-        @build_settings = Helper.backticks(command, print: false)
+        UI.important(command) unless self.xcodebuild_list_silent
```
If we have https://github.com/fastlane/fastlane/pull/5188/files?diff=split#r76524918, we can remove this, right?

KrauseFx commented:
This looks great, thanks for your contribution @fcy. I added some comments that need to be addressed. The comments in the code should be a bit more clear on why this is necessary (maybe link to radar numbers if you have them). 

If you feel like you don't have the time to update the PR, just let me know and I can take over from here 👍 

Thanks again for working on this! :)

KrauseFx commented:
Let me know when you had the time to update the PR and I'll do another review

fcy committed:
Merge branch 'master' into fcy/fix-xcodebuild-show-settings

fcy committed:
Standardize command outputs/prints

- Address PR feedback

fcy commented:
@KrauseFx just pushed the [changes](https://github.com/fastlane/fastlane/pull/5188/commits/c5eb32618aed6bc8b276f891351c925d4b3f1e25). I'll update our `Gemfile` to run it on Monday.

fcy commented on fastlane_core/lib/fastlane_core/project.rb:
```diff
@@ -301,13 +310,42 @@ def self.xcode_list_timeout
+      tries = 1
+      begin
+        Timeout.timeout(timeout) do
+          result = `#{command}`.to_s # Using Helper.backticks doesn't work. Timeout don't times out and the command hangs forever
```
@KrauseFx I don't know what this means 😳

KrauseFx replied:
Oh, the variable is created in the scope of `timeout`, so it won't be available to `return`

KrauseFx replied:
Instead you can just return from here directly, right/

fcy replied:
👍

fcy replied:
Do you prefer the return here? In the new code I created the variable in an outside scope. The only reason I didn't return here was because of the printing the output, I didn't want that to be part of the `retry`.

KrauseFx replied:
I'm ok either way I think

fcy replied:
Also now that I'm thinking about it, does returning from within a closure returns from the method? 🤔 (In Swift it doesn't)

fcy replied:
Ok cool. I'll leave the way it is now, makes more sense to me.

KrauseFx commented:
Hey @fcy, the tests are failing, seems like it's accessing a variable that doesn't exist

fcy commented:
>however please do print out the command and the command output, similar to https://github.com/fastlane/fastlane/blob/master/fastlane_core/lib/fastlane_core/helper.rb#L15-L20

There is a huge addition to the logs by printing the output too, IMO we should just print the command not the output. Check the gist to see the difference: https://gist.github.com/fcy/2cbc57e2876870cee97ac5bf0e5e9dfd

Commit:
Fix undefined variable `result`

Stop printing the command output because it way too verbose. See: https://gist.github.com/fcy/2cbc57e2876870cee97ac5bf0e5e9dfd

Commit:
Merge branch 'master' into fcy/fix-xcodebuild-show-settings

fcy commented:
@KrauseFx fixed the undeclared variable and I went ahead and removed the command output. Let me know if you think we should still print it even though it is really verbose.

Everything is working fine on our CI.

One last request, if when you merge it could you please use Squash and Merge 😁. I'm itching with all the commits in this PR but I decided that not breaking people's `Gemfile.lock` was worth it.

KrauseFx commented:
Thanks. No worries about squashing, we do that 👍

mfurtak commented:
Cool, thanks very much for this contribution! 🚀  Let's merge this, and I'll follow up with a PR to adjust some small details 👍

mfurtak commented:
Hi all, this has been released as part of [`fastlane_core` 0.51.0](https://github.com/fastlane/fastlane/releases/tag/fastlane_core/0.51.0) 🚀 

Thanks very much for your help! Please let us know how it looks 👍

mfurtak commented:
To test the latest release, please run `[sudo] gem update fastlane_core` or if you have a `Gemfile`, `bundle update`

mfurtak commented:
Hi all, [`gym` 1.9.0](https://github.com/fastlane/fastlane/releases/tag/gym/1.9.0) 💪 and [`scan` 0.12.0](https://github.com/fastlane/fastlane/releases/tag/scan/0.12.0) 🔍 have been released to pick up these `fastlane_core` improvements.

To get updated and try them out, please run `[sudo] gem update gym scan` :rocket:
